python/tests/test_tables.py

```diff
@@ -2655,3 +2655,118 @@ def test_empty_nodes(self):
+        pop1 = list(ts.samples(population=0))
+        pop2 = list(ts.samples(population=1))
+        tables1 = ts.simplify(shared_nodes + pop1, record_provenance=False).tables
+        tables2 = ts.simplify(shared_nodes + pop2, record_provenance=False).tables
+        node_mapping = [
+            i if i < len(shared_nodes) else tskit.NULL
+            for i in range(tables2.nodes.num_rows)
+        ]
+        return tables1, tables2, node_mapping
+
+    def get_examples(self, N=5, T=5, seed=242):
+        yield self.split_example(self.get_msprime_example(N, T, seed=seed), T)
+        yield self.split_example(self.get_wf_example(N, T, seed=seed), T)
+
+    def verify_union_equality(self, tables, other, node_mapping, add_populations=True):
+        # verifying against py impl
+        uni1 = tables.copy()
+        uni2 = tables.copy()
+        uni1.union(
+            other,
+            node_mapping,
+            record_provenance=False,
+            add_populations=add_populations,
+        )
+        tsutil.py_union(
+            uni2,
+            other,
+            node_mapping,
+            record_provenance=False,
+            add_populations=add_populations,
+        )
+        self.assertEqual(uni1, uni2)
+        # verifying that subsetting to original nodes return the same table
+        orig_nodes = [j for i, j in enumerate(node_mapping) if j != tskit.NULL]
+        uni1.subset(orig_nodes)
+        # subsetting tables just to make sure order is the same
+        tables.subset(orig_nodes)
+        uni1.provenances.clear()
+        tables.provenances.clear()
+        self.assertEqual(uni1, tables)
+
+    def test_noshared_example(self):
+        ts1 = self.get_msprime_example(sample_size=3, T=2, seed=9328)
+        ts2 = self.get_msprime_example(sample_size=3, T=2, seed=2125)
+        node_mapping = np.full(ts2.num_nodes, tskit.NULL, dtype="int32")
+        uni1 = ts1.union(ts2, node_mapping, record_provenance=False)
+        uni2_tables = ts1.dump_tables()
+        tsutil.py_union(uni2_tables, ts2.tables, node_mapping, record_provenance=False)
+        self.assertEqual(uni1.tables, uni2_tables)
+
+    def test_all_shared_example(self):
+        tables = self.get_wf_example(N=5, T=5, seed=11349).dump_tables()
+        uni = tables.copy()
+        node_mapping = np.arange(tables.nodes.num_rows)
+        uni.union(tables, node_mapping)
+        self.assertEqual(uni.provenances.num_rows, tables.provenances.num_rows + 1)
+        uni.provenances.clear()
+        tables.provenances.clear()
+        self.assertEqual(uni, tables)
+
+    def test_no_add_pop(self):
+        for example in self.get_examples(10, 10, 4357):
+            self.verify_union_equality(*example, add_populations=False)
+
+    def test_examples(self):
+        for N in [2, 4, 5]:
+            for T in [2, 5, 20]:
+                for example in self.get_examples(N, T, 9257):
+                    self.verify_union_equality(*example)
```

Comment:
I'm not seeing any tests for the provenance handling. I think we need to explicitly check this, since it is doing what should be the final version of provenance for the function.